geo/src/algorithm/lines_iter.rs

```diff
@@ -0,0 +1,328 @@
+use crate::{
+    CoordNum, Coordinate, Line, LineString, MultiLineString, MultiPolygon, Polygon, Rect, Triangle,
+};
+use core::slice;
+use std::fmt::Debug;
+use std::iter;
+
+/// Iterate over lines of a geometry.
+pub trait LinesIter<'a> {
+    type Scalar: CoordNum;
+    type Iter: Iterator<Item = Line<Self::Scalar>>;
+
+    /// Iterate over all exterior and (if any) interior lines of a geometry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::{Coordinate, Line};
+    /// use geo::line_string;
+    /// use geo::lines_iter::LinesIter;
+    ///
+    /// let ls = line_string![
+    ///     (x: 1., y: 2.),
+    ///     (x: 23., y: 82.),
+    ///     (x: -1., y: 0.),
+    /// ];
+    ///
+    /// let mut iter = ls.lines_iter();
+    /// assert_eq!(Some(Line::new(Coordinate{x: 1., y: 2.}, Coordinate{x: 23., y:82.})), iter.next());
+    /// assert_eq!(Some(Line::new(Coordinate{x: 23., y: 82.}, Coordinate{x: -1., y:0.})), iter.next());
+    /// assert_eq!(None, iter.next());
+    /// ```
+    fn lines_iter(&'a self) -> Self::Iter;
+}
+
+// ┌────────────────────────────┐
+// │ Implementation for Line    │
+// └────────────────────────────┘
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for Line<T> {
+    type Scalar = T;
+    type Iter = iter::Copied<iter::Once<&'a Line<Self::Scalar>>>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        iter::once(self).copied()
+    }
+}
+
+// ┌──────────────────────────────────┐
+// │ Implementation for LineString    │
+// └──────────────────────────────────┘
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for LineString<T> {
+    type Scalar = T;
+    type Iter = LineStringIter<'a, Self::Scalar>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        LineStringIter::new(self)
+    }
+}
+
+#[doc(hidden)]
```

Comment:
I believe `std` iterator structs aren't `#[doc(hidden)]`, and it's strange to use that for a type returned by the public API.

Reply:
Agreed. I had tried to follow what `coords_iter` does. Added a minimal docstring.